### ISSUE TITLE
M: Week 4 - Tool-calling broken on small local models — supportsTools=false workaround disables web/RAG tools

### ISSUE THREAD

**Issue by Ayyhab** [EduAI-Lab/EduAI#195]
> **Sprint:** Tracked under [#203](https://github.com/EduAI-Lab/EduAI/issues/203). Course/web UX fixes: [#207](https://github.com/EduAI-Lab/EduAI/issues/207), [#208](https://github.com/EduAI-Lab/EduAI/issues/208).  
> **Docs:** [Sprint guide](https://github.com/EduAI-Lab/EduAI/blob/feat/local-models-and-ai-enhancement/docs/team/TEAM_CHAT_LATENCY_SPRINT_GUIDE.md) · [Background](https://github.com/EduAI-Lab/EduAI/blob/feat/local-models-and-ai-enhancement/docs/team/TEAM_CHAT_LATENCY_AND_TOOLS.md)  
> **Code note:** `supportsTools = false` override is **reverted** on `feat/local-models-and-ai-enhancement` — line ~725 uses `await modelSupportsTools(model)` again. Do **not** re-apply the hack.

Part of #61 (EPIC: RAG Pipeline & AI Models).

## Summary

Our chat pipeline registers three tools the model can call to break out of its own training data:

- `getInformation` — RAG search over the selected course's uploaded materials (vector DB).
- `webSearch` — live web search.
- `fetchPage` — download a specific URL.

These exist precisely so the AI is **not limited to its weight memory** (i.e. what the model memorized during training, which has a cutoff date and obviously doesn't contain anything uploaded by our users).

The problem: when we run small local models (e.g. `Llama3.2:latest`, `Gemma3:4b`) through the tool-enabled branch in `apps/core/app/routes/api/chat.ts`, the model often emits a malformed/partial tool call and **stops without producing any user-visible text**. Empirically the request returns:

```
f:{"messageId":"msg-…"}
e:{"finishReason":"stop","usage":{"promptTokens":948,"completionTokens":26}}
d:{"finishReason":"stop","usage":{"promptTokens":948,"completionTokens":26}}
```

26 completion tokens were generated, but **zero text frames (`0:\"…\"`) and zero tool-call frames (`9:…`)** reach the client — the chat bubble renders empty.

## Where it lives

`apps/core/app/routes/api/chat.ts`

- Line 725: `const supportsTools = await modelSupportsTools(model);`
- Lines 866–875: if `supportsTools` is true, `streamText` is called with `tools`, `toolCallStreaming`, and `maxSteps: 3`.

## Current workaround (do NOT ship this)

To unblock local-model demos, we are currently overriding to:

```ts
const supportsTools = false; // changing this for testing purposes from const supportsTools = await modelSupportsTools(model);
```

This forces every request through the hybrid-RAG branch (line 731–829), which:
- Still does a server-side RAG lookup when the user's question contains course-y keywords (\"explain\", \"chapter\", \"summarize\", etc.) and stuffs excerpts into the system prompt.
- **Loses access to `webSearch` and `fetchPage` entirely.** The model can only answer from its own weights + whatever course materials we hand-feed it.

That is a real regression — it caps the AI at \"what was in training data + uploaded files\" and removes the ability to answer questions about current events, external references, links the user pastes, etc. We can't keep this in main.

## What we know about the root cause

- Small Llama / Gemma variants are advertised as tool-capable by their providers, but in practice they emit JSON-shaped output that the AI SDK consumes as a partial tool-call attempt and then drops, leaving an empty response.
- Big frontier models (GPT-4-class, Claude, Gemini Pro, Llama 3.1 70B+) handle the same code path fine.
- This is not a network/streaming bug — `finishReason: \"stop\"` confirms the stream completed normally; the tokens were just never emitted as text.

## Proposed fix direction

Gate tool registration on something stricter than the provider's nominal capability flag. Options worth investigating:

1. **Per-model allowlist** in `~/lib/ai/providers.ts` — `modelSupportsTools` should return `false` for known-broken small models (Llama 3.2 \\<= 3B, Gemma \\<= 9B, etc.) regardless of what the provider advertises.
2. **Size-based gate** — if we can read parameter count from the model metadata, default to tools-off below some threshold.
3. **Per-model config in DB** — admin can toggle tool capability per model from AI Management UI.

Long-term: we also want the hybrid-RAG branch to remain a viable fallback for small models, but **with `webSearch` / `fetchPage` still available** via deterministic keyword routing (server-side decides whether to call them, model just writes the answer with the injected context). That preserves the \"AI is not capped to weight memory\" property even when the model can't do real tool calling.

## Acceptance criteria

- [ ] Remove the `const supportsTools = false;` override from `apps/core/app/routes/api/chat.ts:724`.
- [ ] `modelSupportsTools` returns the correct value per-model (small local models → false; capable models → true).
- [ ] Small local models still have access to `webSearch` / `fetchPage` through a server-driven path (not blocked by tool-calling).
- [ ] Empty-response bug no longer reproduces with Llama 3.2 / Gemma 4B on simple prompts like \"what is gradient descent?\".

Hours to complete: 3 hours [Ayyhab]

**Comment by Ayyhab** [comment]
**Sprint tracking:** Parent epic #203 · Implement server-driven web/RAG fallbacks via #207 #208. 

**Code note (May 2026):** `chat.ts` line ~725 again uses `await modelSupportsTools(model)` — the `supportsTools = false` override described below is **not** on current `feat/local-models-and-ai-enhancement`. Keep this issue for small-model empty-response + per-model tool gating.
